File: src/main/scanner.ts
import fs from 'node:fs'
import path from 'node:path'
import { ItemRecord } from '../shared/types'
import { upsertItems, persistDatabase, getItemByPath } from './db'
import { randomUUID } from 'node:crypto'

/* ============================================================
   Types
   ============================================================ */

interface ScanOptions {
  startPath: string
  mode: 'full' | 'shallow'
  db: any
  dbPath: string
  /** Skip directories already deep-scanned after this ISO date. */
  skipScannedAfter?: string
}

export interface AsyncScanOptions extends ScanOptions {
  /** Called periodically with progress. */
  onProgress?: (info: ScanProgress) => void
  /** If this returns true the scan is aborted. */
  isCancelled?: () => boolean
  /** Optional pre-generated runId. */
  runId?: string
}

export interface ScanProgress {
  runId: string
  itemsScanned: number
  currentPath: string
  state: 'running' | 'completed' | 'error' | 'cancelled'
  message?: string
}

/* ============================================================
   Helpers
   ============================================================ */

/** Returns the real filesystem parent, or null for a root path. */
function fsParent(p: string): string | null {
  const d = path.dirname(p)
  return d === p ? null : d
}

/** Quick stat of a directory's immediate file contents (no recursion). */
function statDirShallow(dirPath: string): {
  sizeBytes: number
  fileCount: number
  folderCount: number
  latestMs: number
} {
  let sizeBytes = 0,
    fileCount = 0,
    folderCount = 0,
    latestMs = 0
  try {
    const entries = fs.readdirSync(dirPath, { withFileTypes: true })
    for (const e of entries) {
      if (e.isFile()) {
        try {
          const s = fs.statSync(path.join(dirPath, e.name))
          sizeBytes += s.size
          fileCount++
          latestMs = Math.max(latestMs, s.mtimeMs)
        } catch {
          /* skip inaccessible */
        }
      } else if (e.isDirectory()) {
        folderCount++
      }
    }
  } catch {
    /* skip inaccessible */
  }
  return { sizeBytes, fileCount, folderCount, latestMs }
}

/** Yield control to the event loop. */
function yieldToEventLoop(): Promise<void> {
  return new Promise((resolve) => setImmediate(resolve))
}

/* ============================================================
   Shallow scan (synchronous — fast enough for a single dir)
   ============================================================ */

function scanShallow(startPath: string, runId: string): ItemRecord[] {
  const root = path.resolve(startPath)
  const items: ItemRecord[] = []
  let entries: fs.Dirent[]
  try {
    entries = fs.readdirSync(root, { withFileTypes: true })
  } catch {
    return items
  }

  let totalSize = 0,
    totalFiles = 0,
    totalFolders = 0,
    latest = 0

  for (const e of entries) {
    const childPath = path.join(root, e.name)
    if (e.isFile()) {
      try {
        const s = fs.statSync(childPath)
        totalSize += s.size
        totalFiles++
        latest = Math.max(latest, s.mtimeMs)
        items.push({
          path: childPath,
          parent: root,
          type: 'File',
          sizeBytes: s.size,
          fileCount: 1,
          folderCount: 0,
          lastWriteUtc: new Date(s.mtimeMs).toISOString(),
          scannedUtc: '',
          depth: 1,
          runId
        })
      } catch {
        continue
      }
    } else if (e.isDirectory()) {
      const di = statDirShallow(childPath)
      totalSize += di.sizeBytes
      totalFolders++
      latest = Math.max(latest, di.latestMs)
      items.push({
        path: childPath,
        parent: root,
        type: 'Folder',
        sizeBytes: di.sizeBytes,
        fileCount: di.fileCount,
        folderCount: di.folderCount,
        lastWriteUtc: new Date(di.latestMs || Date.now()).toISOString(),
        scannedUtc: '',
        depth: 1,
        runId
      })
    }
  }

  // Record the root folder itself
  try {
    const rs = fs.statSync(root)
    latest = Math.max(latest, rs.mtimeMs)
  } catch {
    /* ignore */
  }
  items.push({
    path: root,
    parent: fsParent(root),
    type: 'Folder',
    sizeBytes: totalSize,
    fileCount: totalFiles,
    folderCount: totalFolders,
    lastWriteUtc: new Date(latest || Date.now()).toISOString(),
    scannedUtc: '',
    depth: 0,
    runId
  })

  return items
}

/* ============================================================
   Full recursive scan — async with periodic yielding
   ============================================================ */

interface AggResult {
  sizeBytes: number
  fileCount: number
  folderCount: number
  latestMs: number
}

/** How often (in items) to yield to the event loop & send progress. */
const YIELD_INTERVAL = 200

/** How often (in items) to persist the DB to disk (frees sql.js write buffers). */
const PERSIST_INTERVAL = 5_000

/**
 * Minimum file size (bytes) to store as an individual DB record.
 * Smaller files still count toward their parent folder's totals.
 * This dramatically reduces memory for full-drive scans where millions
 * of tiny files would otherwise bloat the in-memory sql.js database.
 */
const MIN_FILE_SIZE_FOR_DB = 100 * 1024 // 100 KB

/**
 * Async full recursive scan. Yields control to the event loop every
 * YIELD_INTERVAL items so IPC / rendering stays responsive.
 */
async function scanFullAsync(
  dirPath: string,
  depth: number,
  runId: string,
  db: any,
  dbPath: string,
  counter: { count: number; lastYield: number },
  onProgress?: (info: ScanProgress) => void,
  isCancelled?: () => boolean,
  skipScannedAfter?: string
): Promise<AggResult> {
  if (isCancelled?.()) {
    return { sizeBytes: 0, fileCount: 0, folderCount: 0, latestMs: 0 }
  }

  const resolved = path.resolve(dirPath)
  let entries: fs.Dirent[]
  try {
    entries = fs.readdirSync(resolved, { withFileTypes: true })
  } catch {
    return { sizeBytes: 0, fileCount: 0, folderCount: 0, latestMs: 0 }
  }

  const batchItems: ItemRecord[] = []
  let totalSize = 0,
    totalFiles = 0,
    totalFolders = 0,
    latest = 0

  for (const e of entries) {
    if (isCancelled?.()) {
      // Persist what we have so far before bailing out — but do NOT mark
      // this folder as fully scanned (scannedUtc='') since it was cancelled.
      if (batchItems.length > 0) {
        batchItems.push({
          path: resolved,
          parent: fsParent(resolved),
          type: 'Folder',
          sizeBytes: totalSize,
          fileCount: totalFiles,
          folderCount: totalFolders,
          lastWriteUtc: new Date(latest || Date.now()).toISOString(),
          scannedUtc: '',
          depth,
          runId
        })
        upsertItems(db, dbPath, batchItems)
      }
      return { sizeBytes: totalSize, fileCount: totalFiles, folderCount: totalFolders, latestMs: latest }
    }

    const childPath = path.join(resolved, e.name)
    if (e.isFile()) {
      try {
        const s = fs.statSync(childPath)
        totalSize += s.size
        totalFiles++
        latest = Math.max(latest, s.mtimeMs)
        counter.count++
        // Only store files large enough to matter individually
        if (s.size >= MIN_FILE_SIZE_FOR_DB) {
          batchItems.push({
            path: childPath,
            parent: resolved,
            type: 'File',
            sizeBytes: s.size,
            fileCount: 1,
            folderCount: 0,
            lastWriteUtc: new Date(s.mtimeMs).toISOString(),
            scannedUtc: new Date().toISOString(),
            depth: depth + 1,
            runId
          })
        }
      } catch {
        continue
      }
    } else if (e.isDirectory()) {
      // Skip re-scanning directories already scanned after the cutoff
      if (skipScannedAfter) {
        const existing = getItemByPath(db, childPath)
        if (existing && existing.scannedUtc && existing.scannedUtc >= skipScannedAfter) {
          // Use cached values instead of recursing
          totalSize += existing.sizeBytes
          totalFiles += existing.fileCount
          totalFolders += existing.folderCount + 1
          latest = Math.max(latest, new Date(existing.lastWriteUtc).getTime())
          continue
        }
      }

      // Flush batch BEFORE recursing to keep stack-frame memory low
      if (batchItems.length > 0) {
        upsertItems(db, dbPath, batchItems, false)
        batchItems.length = 0
      }

      // Recurse
      const sub = await scanFullAsync(childPath, depth + 1, runId, db, dbPath, counter, onProgress, isCancelled, skipScannedAfter)
      totalSize += sub.sizeBytes
      totalFiles += sub.fileCount
      totalFolders += sub.folderCount + 1
      latest = Math.max(latest, sub.latestMs)
    }

    // Periodically yield to the event loop, flush batch, and send progress
    if (counter.count - counter.lastYield >= YIELD_INTERVAL) {
      counter.lastYield = counter.count
      // Flush accumulated items to free memory
      if (batchItems.length > 0) {
        upsertItems(db, dbPath, batchItems, false)
        batchItems.length = 0
      }
      onProgress?.({
        runId,
        itemsScanned: counter.count,
        currentPath: resolved,
        state: 'running'
      })
      // Periodically persist DB to disk to free sql.js internal write buffers
      if (counter.count % PERSIST_INTERVAL < YIELD_INTERVAL) {
        persistDatabase(db, dbPath)
      }
      await yieldToEventLoop()
    }
  }

  // Record this directory — only mark as scanned if not cancelled
  const wasCancelled = isCancelled?.() ?? false
  try {
    const ds = fs.statSync(resolved)
    latest = Math.max(latest, ds.mtimeMs)
  } catch {
    /* ignore */
  }
  batchItems.push({
    path: resolved,
    parent: fsParent(resolved),
    type: 'Folder',
    sizeBytes: totalSize,
    fileCount: totalFiles,
    folderCount: totalFolders,
    lastWriteUtc: new Date(latest || Date.now()).toISOString(),
    scannedUtc: wasCancelled ? '' : new Date().toISOString(),
    depth,
    runId
  })
  counter.count++

  // Persist this batch (defer disk write — pass false)
  upsertItems(db, dbPath, batchItems, false)

  return { sizeBytes: totalSize, fileCount: totalFiles, folderCount: totalFolders, latestMs: latest }
}

/* ============================================================
   Public API
   ============================================================ */

/** Synchronous scan (shallow only). Returns runId. */
export function runScan({ startPath, mode, db, dbPath }: ScanOptions): string {
  const runId = randomUUID()
  const items = scanShallow(startPath, runId)
  if (items.length > 0) {
    upsertItems(db, dbPath, items)
  }
  return runId
}

/** Active scans that can be cancelled. */
export const activeScans = new Map<string, { cancel: () => void }>()

/** Async scan (full recursive). Returns runId. Yields to event loop periodically. */
export async function runScanAsync({
  startPath,
  mode,
  db,
  dbPath,
  onProgress,
  isCancelled: externalCancel,
  runId: providedRunId,
  skipScannedAfter
}: AsyncScanOptions): Promise<string> {
  const runId = providedRunId ?? randomUUID()

  if (mode === 'shallow') {
    const items = scanShallow(startPath, runId)
    if (items.length > 0) {
      upsertItems(db, dbPath, items)
    }
    onProgress?.({
      runId,
      itemsScanned: items.length,
      currentPath: startPath,
      state: 'completed'
    })
    return runId
  }

  // Full async scan with cancellation support
  let cancelled = false
  activeScans.set(runId, { cancel: () => { cancelled = true } })
  const isCancelled = () => cancelled || (externalCancel?.() ?? false)

  const counter = { count: 0, lastYield: 0 }
  onProgress?.({
    runId,
    itemsScanned: 0,
    currentPath: startPath,
    state: 'running'
  })

  try {
    await scanFullAsync(startPath, 0, runId, db, dbPath, counter, onProgress, isCancelled, skipScannedAfter)

    if (isCancelled()) {
      onProgress?.({
        runId,
        itemsScanned: counter.count,
        currentPath: startPath,
        state: 'cancelled',
        message: `Scan cancelled after ${counter.count} items`
      })
    } else {
      onProgress?.({
        runId,
        itemsScanned: counter.count,
        currentPath: startPath,
        state: 'completed'
      })
    }
  } catch (err: any) {
    onProgress?.({
      runId,
      itemsScanned: counter.count,
      currentPath: startPath,
      state: 'error',
      message: err?.message ?? String(err)
    })
  } finally {
    activeScans.delete(runId)
    // Persist DB to disk once at end of scan
    persistDatabase(db, dbPath)
  }

  return runId
}

File: native/src/scanner.cc
#include "scanner.h"
#include <filesystem>
#include <chrono>
#include <algorithm>
#include <iostream>
#include <thread>

#ifdef _WIN32
#include <windows.h>
#else
#include <sys/stat.h>
#include <dirent.h>
#endif

namespace fs = std::filesystem;

namespace lfb {

Scanner::Scanner() : itemsScanned_(0), lastProgressUpdate_(0) {}
Scanner::~Scanner() {}

static std::string toIsoString(int64_t epochMs) {
    auto tp = std::chrono::system_clock::time_point(std::chrono::milliseconds(epochMs));
    auto time = std::chrono::system_clock::to_time_t(tp);
    std::tm tm;
#ifdef _WIN32
    gmtime_s(&tm, &time);
#else
    gmtime_r(&time, &tm);
#endif
    char buf[32];
    strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", &tm);
    return std::string(buf) + ".000Z";
}

static int64_t fileTimeMs(const fs::file_time_type& ft) {
    auto sctp = std::chrono::time_point_cast<std::chrono::system_clock::duration>(
        ft - fs::file_time_type::clock::now() + std::chrono::system_clock::now()
    );
    return std::chrono::duration_cast<std::chrono::milliseconds>(sctp.time_since_epoch()).count();
}

static std::string getParent(const std::string& path) {
    fs::path p(path);
    if (p.has_parent_path()) {
        auto parent = p.parent_path().string();
        if (parent.empty() || parent == path) return "";
        return parent;
    }
    return "";
}

ScanResult Scanner::statDirShallow(const std::string& dirPath) {
    ScanResult result{0, 0, 0, 0};
    
    try {
        for (const auto& entry : fs::directory_iterator(dirPath, fs::directory_options::skip_permission_denied)) {
            try {
                if (entry.is_regular_file()) {
                    auto size = entry.file_size();
                    result.sizeBytes += size;
                    result.fileCount++;
                    auto lwt = fileTimeMs(entry.last_write_time());
                    result.latestMs = std::max(result.latestMs, lwt);
                } else if (entry.is_directory()) {
                    result.folderCount++;
                }
            } catch (...) {
                // Skip inaccessible entries
            }
        }
    } catch (...) {
        // Skip inaccessible directories
    }
    
    return result;
}

std::vector<ItemRecord> Scanner::scanShallow(const std::string& startPath, const std::string& runId) {
    std::vector<ItemRecord> items;
    
    try {
        int64_t totalSize = 0;
        int totalFiles = 0;
        int totalFolders = 0;
        int64_t latest = 0;

        for (const auto& entry : fs::directory_iterator(startPath, fs::directory_options::skip_permission_denied)) {
            try {
                std::string childPath = entry.path().string();
                
                if (entry.is_regular_file()) {
                    auto size = entry.file_size();
                    auto lwt = fileTimeMs(entry.last_write_time());
                    totalSize += size;
                    totalFiles++;
                    latest = std::max(latest, lwt);
                    
                    ItemRecord rec;
                    rec.path = childPath;
                    rec.parent = startPath;
                    rec.type = "File";
                    rec.sizeBytes = size;
                    rec.fileCount = 1;
                    rec.folderCount = 0;
                    rec.lastWriteUtc = toIsoString(lwt);
                    rec.scannedUtc = "";
                    rec.depth = 1;
                    rec.runId = runId;
                    items.push_back(rec);
                    
                } else if (entry.is_directory()) {
                    auto dirStats = statDirShallow(childPath);
                    totalSize += dirStats.sizeBytes;
                    totalFolders++;
                    latest = std::max(latest, dirStats.latestMs);
                    
                    ItemRecord rec;
                    rec.path = childPath;
                    rec.parent = startPath;
                    rec.type = "Folder";
                    rec.sizeBytes = dirStats.sizeBytes;
                    rec.fileCount = dirStats.fileCount;
                    rec.folderCount = dirStats.folderCount;
                    rec.lastWriteUtc = toIsoString(dirStats.latestMs > 0 ? dirStats.latestMs : 
                        std::chrono::duration_cast<std::chrono::milliseconds>(
                            std::chrono::system_clock::now().time_since_epoch()).count());
                    rec.scannedUtc = "";
                    rec.depth = 1;
                    rec.runId = runId;
                    items.push_back(rec);
                }
            } catch (...) {
                continue;
            }
        }
        
        // Add root folder record
        try {
            auto rootEntry = fs::directory_entry(startPath);
            if (rootEntry.exists()) {
                auto rootLwt = fileTimeMs(rootEntry.last_write_time());
                latest = std::max(latest, rootLwt);
            }
        } catch (...) {}
        
        ItemRecord rootRec;
        rootRec.path = startPath;
        rootRec.parent = getParent(startPath);
        rootRec.type = "Folder";
        rootRec.sizeBytes = totalSize;
        rootRec.fileCount = totalFiles;
        rootRec.folderCount = totalFolders;
        rootRec.lastWriteUtc = toIsoString(latest > 0 ? latest : 
            std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::system_clock::now().time_since_epoch()).count());
        rootRec.scannedUtc = "";
        rootRec.depth = 0;
        rootRec.runId = runId;
        items.push_back(rootRec);
        
    } catch (...) {
        // Return whatever we collected
    }
    
    return items;
}

ScanResult Scanner::scanFullAsync(
    const std::string& startPath,
    int depth,
    const std::string& runId,
    std::atomic<bool>& cancelled,
    ProgressCallback progressCb,
    const std::string& skipScannedAfter
) {
    ScanResult result{0, 0, 0, 0};
    
    if (cancelled.load()) {
        return result;
    }
    
    try {
        std::vector<ItemRecord> batchItems;
        
        for (const auto& entry : fs::directory_iterator(startPath, fs::directory_options::skip_permission_denied)) {
            if (cancelled.load()) {
                break;
            }
            
            try {
                std::string childPath = entry.path().string();
                
                if (entry.is_regular_file()) {
                    auto size = entry.file_size();
                    auto lwt = fileTimeMs(entry.last_write_time());
                    result.sizeBytes += size;
                    result.fileCount++;
                    result.latestMs = std::max(result.latestMs, lwt);
                    
                    itemsScanned_++;
                    
                    // Only store large files
                    if (size >= MIN_FILE_SIZE_FOR_DB) {
                        ItemRecord rec;
                        rec.path = childPath;
                        rec.parent = startPath;
                        rec.type = "File";
                        rec.sizeBytes = size;
                        rec.fileCount = 1;
                        rec.folderCount = 0;
                        rec.lastWriteUtc = toIsoString(lwt);
                        rec.scannedUtc = toIsoString(std::chrono::duration_cast<std::chrono::milliseconds>(
                            std::chrono::system_clock::now().time_since_epoch()).count());
                        rec.depth = depth + 1;
                        rec.runId = runId;
                        batchItems.push_back(rec);
                    }
                    
                } else if (entry.is_directory()) {
                    // TODO: Check skipScannedAfter logic with DB
                    
                    // Recurse
                    auto subResult = scanFullAsync(childPath, depth + 1, runId, cancelled, progressCb, skipScannedAfter);
                    result.sizeBytes += subResult.sizeBytes;
                    result.fileCount += subResult.fileCount;
                    result.folderCount += subResult.folderCount + 1;
                    result.latestMs = std::max(result.latestMs, subResult.latestMs);
                }
                
                // Progress reporting
                int current = itemsScanned_.load();
                int lastUpdate = lastProgressUpdate_.load();
                if (current - lastUpdate >= PROGRESS_INTERVAL) {
                    if (lastProgressUpdate_.compare_exchange_strong(lastUpdate, current)) {
                        ScanProgress progress;
                        progress.runId = runId;
                        progress.itemsScanned = current;
                        progress.currentPath = startPath;
                        progress.state = "running";
                        if (progressCb) {
                            progressCb(progress);
                        }
                    }
                }
                
            } catch (...) {
                continue;
            }
        }
        
        // TODO: Batch persist to database
        // For now, just track items in memory
        
    } catch (...) {
        // Return partial result
    }
    
    return result;
}

} // namespace lfb
